File: examples/step06/pkg/timer/Timer.py
# -*- Python -*-
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2026 all rights reserved

# support
import time


class Timer:
    """
    A wall-clock timer backed by time.perf_counter.
    """

    def __init__(self, **kwds):
        # chain
        super().__init__(**kwds)
        # state
        self._running: bool = False
        self._start: float = 0.0
        self._elapsed: float = 0.0

    def start(self) -> "Timer":
        """
        Start the timer; no-op if already running.
        """
        if not self._running:
            self._start = time.perf_counter()
            self._running = True
        return self

    def stop(self) -> "Timer":
        """
        Stop the timer and accumulate elapsed time; no-op if not running.
        """
        if self._running:
            self._elapsed += time.perf_counter() - self._start
            self._running = False
        return self

    def reset(self) -> "Timer":
        """
        Clear the accumulated time and stop the timer.
        """
        self._running = False
        self._elapsed = 0.0
        return self

    @property
    def elapsed(self) -> float:
        """
        The accumulated time in seconds, including any currently open interval.
        """
        total = self._elapsed
        if self._running:
            total += time.perf_counter() - self._start
        return total

    @property
    def running(self) -> bool:
        """
        Whether the timer is currently running.
        """
        return self._running


# end of file

File: examples/step03/ext/timer/Timer.cc
// -*- C++ -*-
// -*- coding: utf-8 -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2026 all rights reserved

// external
#include "external.h"
// namespace setup
#include "forward.h"


// bind the {Timer} class
void
timer::py::Timer(py::module & m)
{
    // type alias
    using timer_t = timer::Timer;

    // create the class
    auto cls = py::class_<timer_t>(m, "Timer");

    // constructors
    cls.def(
        // the default constructor
        py::init<>());

    // interface
    cls.def(
        // the name
        "start",
        // the implementation; discard the C++ reference return so Python gets None
        [](timer_t & t) { t.start(); },
        // the docstring
        "start the timer; no-op if already running");

    cls.def(
        "stop", [](timer_t & t) { t.stop(); },
        "stop the timer and accumulate elapsed time; no-op if not running");

    cls.def(
        "reset", [](timer_t & t) { t.reset(); }, "clear the accumulated time and stop the timer");

    // read-only properties
    cls.def_property_readonly(
        // the name
        "elapsed",
        // the getter
        &timer_t::elapsed,
        // the docstring
        "accumulated time in seconds, including any open interval");

    cls.def_property_readonly(
        "running", &timer_t::running, "whether the timer is currently running");

    // all done
    return;
}


// end of file
